src/payment/strategies/mollie/mollie-payment-strategy.ts

```diff
@@ -71,16 +85,29 @@ export default class MolliePaymentStrategy implements PaymentStrategy {
         const shouldSaveInstrument = (paymentData as HostedInstrument)?.shouldSaveInstrument;
         const shouldSetAsDefaultInstrument = (paymentData as HostedInstrument)?.shouldSetAsDefaultInstrument;
 
-        if (!payment || !paymentData) {
-            throw new PaymentArgumentInvalidError([ 'payment' ]);
+        if (!payment || !payment.gatewayId || !paymentData) {
+            throw new PaymentArgumentInvalidError([ 'payment', 'gatewayId', 'paymentData' ]);
         }
 
         try {
             if (payment.methodId === MolliePaymentMethodType.creditcard) {
                 await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
 
                 if (paymentData && isVaultedInstrument(paymentData)) {
-                    return this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
+                    if (this._isHostedPaymentFormEnabled(payment.methodId, payment.gatewayId) && this._shouldRenderHostedForm) {
+                        const form = this._hostedForm;
+
+                        if (!form) {
+                            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                        }
+
+                        await form.validate();
+                        await form.submit(payment);
+
+                        return this._store.dispatch(this._orderActionCreator.loadCurrentOrder());
+                    } else {
+                        return await this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
```

Comment:
How come we `await` for the last `distpatch` but not the first one? Do we need to `await` for any of them... 🤷